### PR TITLE
Persistent Node Ids

### PR DESCRIPTION
**Pull request by bleskes**
Node IDs are currently randomly generated during node startup. That means they change every time the node is restarted. While this doesn't matter for ES proper, it makes it hard for external services to track nodes. Another, more minor, side effect is that indexing the output of, say, the node stats API results in creating new fields due to node ID being used as keys.

The first approach I considered was to use the node's published address as the base for the id. We already [treat nodes with the same address as the same](https://github.com/elastic/elasticsearch/blob/master/core/src/main/java/org/elasticsearch/discovery/zen/NodeJoinController.java#L387) so this is a simple change (see [here](https://github.com/elastic/elasticsearch/compare/master...bleskes:node_persistent_id_based_on_address)). While this is simple and it works for probably most cases, it is not perfect. For example, if after a node restart, the node is not able to bind to the same port (because it's not yet freed by the OS), it will cause the node to still change identity. Also in environments where the host IP can change due to a host restart, identity will not be the same. 

Due to those limitation, I opted to go with a different approach where the node id will be persisted in the node's data folder. This has the upside of connecting the id to the nodes data. It also means that the host can be adapted in any way (replace network cards, attach storage to a new VM). I

It does however also have downsides - we now run the risk of two nodes having the same id, if someone copies clones a data folder from one node to another. To mitigate this I changed the semantics of the protection against multiple nodes with the same address to be stricter - it will now reject the incoming join if a node exists with the same id but a different address. Note that if the existing node doesn't respond to pings (i.e., it's not alive) it will be removed and the new node will be accepted when it tries another join.

Last, and most importantly, this change requires that _all_ nodes persist data to disk. This is a change from current behavior where only data & master nodes store local files. This is the main reason for marking this PR as breaking.

Other less important notes:
- DummyTransportAddress is removed as we need a unique network address per node. Use `LocalTransportAddress.buildUnique()` instead.
- I renamed `node.add_lid_to_custom_path` to `node.add_lock_id_to_custom_path` to avoid confusion with the node ID which is now part of the `NodeEnvironment` logic.
- I removed the `version` paramater from `MetaDataStateFormat#write` , it wasn't really used and was just in the way :)
- TribeNodes are special in the sense that they do start multiple sub-nodes (previously known as client nodes). Those sub-nodes do not store local files but derive their ID from the parent node id, so they are generated consistently.

This PR supersedes #17811, and changes it by adding an ephimeralID field to DiscoveryNode that maintains the current id semantics, i.e., it changes with each node restart. This allows to keep the same semantics and use it for node equality.

### CODE REVIEW

File: core/src/main/java/org/elasticsearch/env/NodeEnvironment.java [hunk 1]
```diff
@@ -360,6 +376,28 @@ private void maybeLogHeapDetails() {
...
 
 
+    /**
+     * scans the node paths and loads existing metaData file. If not found a new meta data will be generated
+     * and persisted into the nodePaths
+     */
+    private static NodeMetaData loadOrCreateNodeMetaData(Settings settings, ESLogger logger,
+                                                         NodePath... nodePaths) throws IOException {
+        List<Path> pathList = Arrays.stream(nodePaths).map(np -> np.path).collect(Collectors.toList());
+        final Path[] paths = pathList.toArray(new Path[pathList.size()]);
```
Comment (ywelsch) [inline core/src/main/java/org/elasticsearch/env/NodeEnvironment.java]: The above two lines can be simplified to
`final Path[] paths = Arrays.stream(nodePaths).map(np -> np.path).toArray(Path[]::new);`

File: core/src/main/java/org/elasticsearch/cluster/node/DiscoveryNode.java [hunk 2]
```diff
@@ -238,6 +251,13 @@ public String getId() {
     }
 
     /**
+     * The unique ephemeral id of the node.
+     */
+    public String getEphemeralId() {
```
Comment (ywelsch) [inline core/src/main/java/org/elasticsearch/cluster/node/DiscoveryNode.java]: see https://github.com/elastic/elasticsearch/pull/17987/files#r63505561
can we add some docs as to what we mean with ephemeral - most notably when it changes, and it's implication to identity?

File: core/src/main/java/org/elasticsearch/cluster/node/DiscoveryNode.java [hunk 3]
```diff
@@ -97,14 +106,15 @@ public static boolean isIngestNode(Settings settings) {
      * and updated.
      * </p>
      *
-     * @param nodeId     the nodes unique id.
-     * @param address    the nodes transport address
-     * @param attributes node attributes
-     * @param roles      node roles
-     * @param version    the version of the node.
+     * @param id               the nodes unique (ephemeral and persistent) node id
```
Comment (ywelsch) [inline core/src/main/java/org/elasticsearch/cluster/node/DiscoveryNode.java]: this isn't true anymore. The ephemeral id is randomly generated when using this constructor.

Comment (ywelsch) [inline core/src/main/java/org/elasticsearch/cluster/node/DiscoveryNode.java]: In general I find these overloaded constructors dangerous, making it highly likely that some places are using the wrong one or unaware of the implications of generating a fresh ephemeral id.

File: core/src/main/java/org/elasticsearch/cluster/node/DiscoveryNodes.java [hunk 4]
```diff
@@ -163,6 +163,18 @@ public boolean nodeExists(String nodeId) {
     }
 
     /**
+     * Determine if a given node exists
+     *
+     * @param node of the node which existence should be verified
+     * @return <code>true</code> if the node exists. Otherwise <code>false</code>
+     */
+    public boolean nodeExists(DiscoveryNode node) {
```
Comment (ywelsch) [inline core/src/main/java/org/elasticsearch/cluster/node/DiscoveryNodes.java]: I see two usages of `nodeExists(String nodeId)` that could use this one instead. (LocalDiscovery and IndicesClusterStateServiceRandomUpdatesTests)

File: core/src/main/java/org/elasticsearch/cluster/node/DiscoveryNodes.java [hunk 5]
```diff
@@ -592,6 +630,30 @@ public Builder localNodeId(String localNodeId) {
...
 
+        /**
+         * Checks that a node can be safely added to this node collection.
+         *
+         * @return null if all is OK or an error message explaining why a node can not be added.
+         *
+         * Note: if this method returns a non-null value, calling {@link #put(DiscoveryNode)} will fail with an
+         * exception
+         */
+        public String preflightPut(DiscoveryNode node) {
```
Comment (ywelsch) [inline core/src/main/java/org/elasticsearch/cluster/node/DiscoveryNodes.java]: `preflight` is the weirdest word and does not convey anything meaningful to me in the context of this class. Maybe call it `checkSafeAdd`?

Comment (ywelsch) [inline core/src/main/java/org/elasticsearch/cluster/node/DiscoveryNodes.java]: also this method can be private now.

File: core/src/main/java/org/elasticsearch/discovery/local/LocalDiscovery.java [hunk 6]
```diff
@@ -102,6 +103,13 @@ public void startInitialJoin() {
             }
             logger.debug("Connected to cluster [{}]", clusterName);
 
+            Optional<LocalDiscovery> current = clusterGroup.members().stream().filter(other -> other.localNode().equals(this.localNode()))
+                .findFirst();
+            if (current.isPresent()) {
+                throw new IllegalStateException("current cluster group already contains a node with the same id. current "
```
Comment (ywelsch) [inline core/src/main/java/org/elasticsearch/discovery/local/LocalDiscovery.java]: This only checks that it does not have the same ephemeral id, not node id?

File: core/src/main/java/org/elasticsearch/discovery/zen/NodeJoinController.java [hunk 7]
```diff
@@ -431,16 +432,15 @@ public void clusterStateProcessed(String source, ClusterState oldState, ClusterS
...
-                    for (DiscoveryNode existingNode : currentNodes) {
-                        if (node.getAddress().equals(existingNode.getAddress())) {
-                            nodesBuilder.remove(existingNode.getId());
-                            logger.warn("received join request from node [{}], but found existing node {} with same address, removing existing node", node, existingNode);
-                        }
+                    try {
+                        nodesBuilder.put(node);
+                        nodesChanged = true;
+                    } catch (IllegalArgumentException e) {
+                        results.failure(node, e);
```
Comment (ywelsch) [inline core/src/main/java/org/elasticsearch/discovery/zen/NodeJoinController.java]: We should do some debug logging here.

File: core/src/main/java/org/elasticsearch/discovery/zen/fd/NodesFaultDetection.java [hunk 8]
```diff
@@ -275,8 +280,8 @@ public void messageReceived(PingRequest request, TransportChannel channel) throw
 
     public static class PingRequest extends TransportRequest {
 
-        // the (assumed) node id we are pinging
-        private String nodeId;
+        // the (assumed) node we are pinging
+        private DiscoveryNode node;
```
Comment (ywelsch) [inline core/src/main/java/org/elasticsearch/discovery/zen/fd/NodesFaultDetection.java]: maybe call this `targetNode` so it is clearer that it's the node we are pinging?

File: core/src/main/java/org/elasticsearch/discovery/zen/fd/NodesFaultDetection.java [hunk 9]
```diff
@@ -256,14 +260,15 @@ public String executor() {
         public void messageReceived(PingRequest request, TransportChannel channel) throws Exception {
             // if we are not the node we are supposed to be pinged, send an exception
             // this can happen when a kill -9 is sent, and another node is started using the same port
-            if (!localNode.getId().equals(request.nodeId)) {
-                throw new IllegalStateException("Got pinged as node [" + request.nodeId + "], but I am node [" + localNode.getId() + "]");
+            if (!localNode.equals(request.node())) {
+                throw new IllegalStateException("Got pinged as node " + request.node() + "], but I am node " + localNode );
```
Comment (ywelsch) [inline core/src/main/java/org/elasticsearch/discovery/zen/fd/NodesFaultDetection.java]: I like that these exceptions are going to be much more useful with all node information contained (no more cryptic ids).

File: core/src/main/java/org/elasticsearch/env/NodeEnvironment.java [hunk 10]
```diff
@@ -134,23 +140,30 @@ public String toString() {
...
-     * If true automatically append node id to custom data paths.
+     * If true automatically append node lock id to custom data paths.
      */
-    public static final Setting<Boolean> ADD_NODE_ID_TO_CUSTOM_PATH =
-        Setting.boolSetting("node.add_id_to_custom_path", true, Property.NodeScope);
+    public static final Setting<Boolean> ADD_NODE_LOCK_ID_TO_CUSTOM_PATH =
+        Setting.boolSetting("node.add_lock_id_to_custom_path", true, Property.NodeScope);
+
+
+    public static final Setting<Long> NODE_ID_SEED_SETTING =
```
Comment (ywelsch) [inline core/src/main/java/org/elasticsearch/env/NodeEnvironment.java]: add some javadocs here:

From #17987:
```
/**
  * Seed for determining the unique uuid of this node. If the node has already a persisted uuid on disk,
  * this seed will be ignored and the uuid from disk will be reused.
  */
```

File: core/src/main/java/org/elasticsearch/env/NodeMetaData.java [hunk 11]
```diff
@@ -0,0 +1,114 @@
...
+import org.elasticsearch.common.xcontent.XContentParser;
+import org.elasticsearch.common.xcontent.XContentType;
+import org.elasticsearch.gateway.MetaDataStateFormat;
+
+import java.io.IOException;
+import java.io.OutputStream;
+import java.util.Objects;
+
+/**
+ */
```
Comment (ywelsch) [inline core/src/main/java/org/elasticsearch/env/NodeMetaData.java]: see https://github.com/elastic/elasticsearch/pull/17987/files#r61740801

File: core/src/main/java/org/elasticsearch/gateway/MetaDataStateFormat.java [hunk 12]
```diff
@@ -182,13 +183,18 @@ public final T read(Path file) throws IOException {
             try (final IndexInput indexInput = dir.openInput(file.getFileName().toString(), IOContext.DEFAULT)) {
                  // We checksum the entire file before we even go and parse it. If it's corrupted we barf right here.
                 CodecUtil.checksumEntireFile(indexInput);
-                CodecUtil.checkHeader(indexInput, STATE_FILE_CODEC, STATE_FILE_VERSION, STATE_FILE_VERSION);
+                final int fileVersion = CodecUtil.checkHeader(indexInput, STATE_FILE_CODEC, MIN_COMPATIBLE_STATE_FILE_VERSION,
+                    STATE_FILE_VERSION);
                 final XContentType xContentType = XContentType.values()[indexInput.readInt()];
-                indexInput.readLong(); // version currently unused
+                if (fileVersion == 0) {
```
Comment (ywelsch) [inline core/src/main/java/org/elasticsearch/gateway/MetaDataStateFormat.java]: magic constant. would be useful to know which ES versions used this constant, see https://github.com/elastic/elasticsearch/pull/17987/files#diff-35f0f7b57ab6d12add07c462b761253bR67

File: core/src/test/java/org/elasticsearch/cluster/node/DiscoveryNodesTests.java [hunk 13]
```diff
@@ -98,26 +104,98 @@ public void testResolveNodesIds() {
...
+        Set<DiscoveryNode> nodesA = new HashSet<>();
+        nodesA.addAll(randomNodes(1 + randomInt(10)));
+        Set<DiscoveryNode> nodesB = new HashSet<>();
+        nodesB.addAll(randomNodes(1 + randomInt(5)));
+        for (DiscoveryNode node : randomSubsetOf(nodesA)) {
+            if (randomBoolean()) {
+                // change an attribute
+                Map<String, String> attrs = new HashMap<>(node.getAttributes());
+                attrs.put("new", "new");
+                node = new DiscoveryNode(node.getName(), node.getId(), node.getAddress(), attrs, node.getRoles(), node.getVersion());
```
Comment (ywelsch) [inline core/src/test/java/org/elasticsearch/cluster/node/DiscoveryNodesTests.java]: here you also generate a new ephemeral id, probably not intended...

**Comment by ywelsch** [comment]
@bleskes Left some comments on the PR but looks good overall. Can you add documentation as well? (breaking change). I'm ok if you want to address tribe nodes as well in this PR (I don't think it's too big of a change, see #17987).

File: core/src/main/java/org/elasticsearch/env/NodeEnvironment.java (hunk 1 above)
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/env/NodeEnvironment.java]: Agreed.

File: core/src/main/java/org/elasticsearch/cluster/node/DiscoveryNode.java (hunk 2 above)
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/cluster/node/DiscoveryNode.java]: :) good idea..

File: core/src/main/java/org/elasticsearch/cluster/node/DiscoveryNode.java (hunk 3 above)
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/cluster/node/DiscoveryNode.java]: I found it very cumbersome to generates those uuids in all the tests. I think that if you don't specify an ephemeral id, it's safe to assume you want a new node? if you don't and you need equality you will find out soon enough?

File: core/src/main/java/org/elasticsearch/cluster/node/DiscoveryNodes.java (hunk 4 above)
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/cluster/node/DiscoveryNodes.java]: good catches. fixed.

File: core/src/main/java/org/elasticsearch/cluster/node/DiscoveryNodes.java (hunk 5 above)
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/cluster/node/DiscoveryNodes.java]: I renamed this to `validatePut` and made it private

File: core/src/main/java/org/elasticsearch/discovery/local/LocalDiscovery.java (hunk 6 above)
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/discovery/local/LocalDiscovery.java]: added a check for node id as well.

File: core/src/main/java/org/elasticsearch/discovery/zen/NodeJoinController.java (hunk 7 above)
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/discovery/zen/NodeJoinController.java]: we get those for free from : https://github.com/elastic/elasticsearch/blob/master/core/src/main/java/org/elasticsearch/cluster/service/ClusterService.java#L585

File: core/src/main/java/org/elasticsearch/discovery/zen/fd/NodesFaultDetection.java (hunk 8 above)
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/discovery/zen/fd/NodesFaultDetection.java]: done.

File: core/src/main/java/org/elasticsearch/env/NodeEnvironment.java (hunk 10 above)
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/env/NodeEnvironment.java]: added

File: core/src/main/java/org/elasticsearch/env/NodeMetaData.java (hunk 11 above)
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/env/NodeMetaData.java]: copied the java docs over. Thx.

File: core/src/main/java/org/elasticsearch/gateway/MetaDataStateFormat.java (hunk 12 above)
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/gateway/MetaDataStateFormat.java]: Added.

File: core/src/test/java/org/elasticsearch/cluster/node/DiscoveryNodesTests.java (hunk 13 above)
Comment (bleskes) [inline core/src/test/java/org/elasticsearch/cluster/node/DiscoveryNodesTests.java]: it actually is - the addition of new attributes means the node had to be restarted and got a new ephemeral id.

**Comment by bleskes** [comment]
@ywelsch Thanks. I pushed a commit addressing your comments. I'm not sure about the tribe nodes - will reach out to discuss more.

**Comment by bleskes** [comment]
@ywelsch I pushed another commit with the tribe node change we discussed.

File: core/src/test/java/org/elasticsearch/cluster/node/DiscoveryNodesTests.java (hunk 13 above)
Comment (ywelsch) [inline core/src/test/java/org/elasticsearch/cluster/node/DiscoveryNodesTests.java]: right!

**Comment by ywelsch** [comment]
LGTM. Can you also add something to the migration docs? (see e.g. https://github.com/elastic/elasticsearch/pull/17987/files#diff-2d50fb3821a6a54aedf97e29135d711aR13 )

**Pull request merged**